soroban-env-host/src/host/prng.rs

```diff
@@ -0,0 +1,151 @@
+use crate::{
+    budget::Budget,
+    host::metered_clone::MeteredClone,
+    host_object::HostVec,
+    xdr::{ContractCostType, ScBytes},
+    HostError,
+};
+use rand::{distributions::Uniform, prelude::Distribution, seq::SliceRandom};
+use rand_chacha::{
+    rand_core::{RngCore, SeedableRng},
+    ChaCha20Rng,
+};
+use std::ops::RangeInclusive;
+
+/// PRNG subsystem in the host, which provides best-effort pseudo-randomness to
+/// guest contracts using a combination of features that guests cannot easily
+/// implement themselves:
```

Comment:
just to emphasize: we aren't (shouldn't!) recommend this for cryptographic usage in any way, shape, or form, right?

Reply:
It depends what you mean by "cryptographic uses". ChaCha20 is a CSPRNG (in terms of its mixing function) and I don't think we need to shy away from committing to providing one. On the other hand it has to operate (a) in consensus among all peers and (b) using strictly public input data, so the main thing people _usually_ want out of a CSPRNG and usually mean when they say "cryptographic usage" -- non-public, secret state that's usable to derive secret keys or unpredictable nonces from, that are then not-shared -- is by definition not available. Every other node in the system (including watchers) can immediately calculate / observe the exact internal state of the ChaCha20 instance by running their own copy on the same input that everyone has, and everything you derive from the state during execution is by necessity public and identical to the copies everyone else calculates in consensus.

The point of this is to allow contracts to "make unpredictable choices" in their actions that are as _close_ to unpredictable-in-advance as we can get them, given this by-defnition public-and-predictable context. An example might be a lottery contract that's going to pay out to a single participant. In theory this should be done by having randomness provided by an off-chain oracle (or set of oracles) that are trusted not to be participating in the lottery, and that first submit commitments and then later submit signed preimages. But that's a lot of work to set up and still relies on a trusted third party, and so people often (especially in the early days of a given contract)  don't do it. Instead they often build "bad randomness sources" like the ledger sequence number, which is _very_ predictable in advance. Someone wanting to win a lottery would be able to guess the ledger sequence number of the payout ledger in advance (and/or as soon as they see that a ledger is closing with a "do the payout" transaction), and could use that knowledge to situate themselves as the winner.

In the design here (and in the corresponding patch we'll land on stellar-core) the seed is intended to be derived from the txset that won nomination, and the tx's hash-order numerical position within it, which is both hard to predict and by definition changes if you observe it and then inject your own transaction. In general it's just hard to influence unless the attacker is themselves a trusted validator, since nominations from anyone else are ignored. And the network generally already affords validators this level of trust (they can already reorder or drop transactions during consensus -- or inject their own -- when picking a txset to nominate, so we're not really increasing the MEV attack surface here).